### PR TITLE
Cody Recipe Widget for Code Snippet Responses

### PR DESCRIPTION
**Pull request by thenamankumar**
Closes [#55189
](https://github.com/sourcegraph/sourcegraph/issues/55189). Paired on with @st0nebraker and @thenamankumar.

Added props and refs needed for cody tooltip to be available on code snippets from cody chat.

Merge this before [#55365](https://github.com/sourcegraph/sourcegraph/pull/55365)

### CODE REVIEW

File: lib/ui/src/chat/CodeBlocks.tsx [hunk 1]
```diff
@@ -8,13 +8,19 @@ import { CopyButtonProps } from '../Chat'
 
 import styles from './CodeBlocks.module.css'
 
+export type CodyRecipesWidgetWrapper = React.ComponentType<{
+    targetRef: RefObject<HTMLElement>
+    children: React.ReactNode
+}> | null
```
Comment (courier-new) [inline lib/ui/src/chat/CodeBlocks.tsx:14]: Nit: If the prop is optional, what if we simplify this type to always be non-null:

```suggestion
export type CodyRecipesWidgetWrapper = React.ComponentType<{
    targetRef: RefObject<HTMLElement>
    children: React.ReactNode
}>
```

And then just assign a default value of `null` in the component prop destructuring?

File: lib/ui/src/chat/CodeBlocks.tsx [hunk 2]
```diff
@@ -97,6 +103,7 @@ export const CodeBlocks: React.FunctionComponent<CodeBlocksProps> = React.memo(f
     copyButtonClassName,
     insertButtonClassName,
     CopyButtonProps,
+    RecipesWidgetWrapper,
```
Comment (courier-new) [inline lib/ui/src/chat/CodeBlocks.tsx:106]: Per the other suggestion,

```suggestion
    RecipesWidgetWrapper = null,
```

File: lib/ui/src/chat/CodeBlocks.tsx [hunk 3]
```diff
@@ -119,8 +126,25 @@ export const CodeBlocks: React.FunctionComponent<CodeBlocksProps> = React.memo(f
...
-        () => <div ref={rootRef} dangerouslySetInnerHTML={{ __html: renderCodyMarkdown(displayText) }} />,
-        [displayText]
+    const RecipesWidgetWrapperWithProps = useMemo(
+        () => (props: { targetRef: RefObject<HTMLElement>; children: React.ReactNode }) => {
+            if (RecipesWidgetWrapper) {
+                return <RecipesWidgetWrapper {...props} />
+            }
+            return null
+        },
+        [rootRef]
```
Comment (courier-new) [inline lib/ui/src/chat/CodeBlocks.tsx:136]: It's not immediately obvious to me why this memoized value has a dependency on `rootRef`, and not `RecipesWidgetWrapper`? 🤔

I'm also not 100% clear on why this needs to be memoized in the first place, if we're also memoizing the whole return value of this component, *and* the component itself is `React.memo`ized. But maybe you have lots of reasons! I'd love to understand them, and perhaps also document them inline so it's easier for someone else to understand, too. 🙂

**Review by courier-new — APPROVED** [review]
Looks good, just had a couple small suggestions and questions!

File: lib/ui/src/chat/CodeBlocks.tsx (hunk 3 above)
Comment (st0nebraker) [reply lib/ui/src/chat/CodeBlocks.tsx]: I tried this without the `React.memo` and there was quite a delay in the tooltip from unmounting, so it appears it is helpful here

**Pull request closed without merging**